### PR TITLE
Fixed #28766 -- Added ResolverMatch.route

### PR DESCRIPTION
**Pull request by xavfernandez**
https://code.djangoproject.com/ticket/28766

### CODE REVIEW

File: docs/ref/urlresolvers.txt [hunk 1]
```diff
@@ -130,6 +130,12 @@ If the URL does not resolve, the function raises a
 
         The name of the URL pattern that matches the URL.
 
+    .. attribute:: ResolverMatch.route
+
+        .. versionadded:: 2.1
```
Comment (auvipy) [inline docs/ref/urlresolvers.txt]: I believe this should be 2.2?

File: docs/ref/urlresolvers.txt [hunk 2]
```diff
@@ -130,6 +130,12 @@ If the URL does not resolve, the function raises a
...
+
+        The route of the URL pattern that matches the URL.
```
Comment (auvipy) [inline docs/ref/urlresolvers.txt]: any example would be a nice addition to the proposed change!

File: docs/ref/urlresolvers.txt (hunk 1 above)
Comment (timgraham) [reply docs/ref/urlresolvers.txt]: Yes, and please add a release note in `docs/releases/2.2.txt`.

**Comment by timgraham** [comment]
Please check test coverage carefully. I don't see any failures if the change in `ResolverMatch.__repr__()` is reverted, for example. Uncheck "patch needs improvement" on the ticket when it's ready for review again.

File: tests/urlpatterns/tests.py [hunk 3]
```diff
@@ -26,23 +26,38 @@ def test_path_lookup_without_parameters(self):
...
+    def test_re_path(self):
+        match = resolve('/regex/1/')
+        self.assertEqual(match.url_name, 'regex')
+        self.assertEqual(match.kwargs, {'pk': '1'})
+        self.assertEqual(match.route, '^regex/(?P<pk>[0-9]+)/$')
+
+    def test_path_lookup_with_inclusion(self):
+        match = resolve('/included_urls/extra/something/')
+        self.assertEqual(match.url_name, 'inner-extra')
+        self.assertEqual(match.route, r'included_urls/^extra/(?P<extra>\w+)/$')
```
Comment (xavfernandez) [inline tests/urlpatterns/tests.py]: I'm wondering whether we want  `r'included_urls/^extra/(?P<extra>\w+)/$` or `r'included_urls/extra/(?P<extra>\w+)/$` (with or without the middle `^`). Any opinion ?

Comment (mes3yd) [reply tests/urlpatterns/tests.py]: Voting without: It's debugging sugar, not a valid route, so we have no way to match incoming requests with a route without stripping it again.
It would be better to have a property `route_components` that would return a sequence of paths/routes that one could join with whatever separator one wants to show the inclusion.

Comment (timgraham) [reply tests/urlpatterns/tests.py]: Without makes sense to me.

Comment (xavfernandez) [reply tests/urlpatterns/tests.py]: I removed it.

**Comment by xavfernandez** [comment]
Hello @timgraham, sorry to nag but is there anything missing or something I must/could do to help this get merged ?

File: django/urls/resolvers.py [hunk 4]
```diff
@@ -317,6 +318,9 @@ def __init__(self, pattern, callback, default_args=None, name=None):
         self.default_args = default_args or {}
         self.name = name
 
+    # def __str__(self):
```
Comment (timgraham) [inline django/urls/resolvers.py]: A test is missing to show why this change belongs in the PR.

File: tests/urlpatterns/path_urls.py [hunk 5]
```diff
@@ -11,5 +11,7 @@
     path('users/', views.empty_view, name='users'),
     path('users/<id>/', views.empty_view, name='user-with-id'),
     path('included_urls/', include('urlpatterns_reverse.included_urls')),
+    re_path(r'^regex/(?P<pk>[0-9]+)/$', views.empty_view, name='regex'),
+    path('', include('urlpatterns_reverse.extra_urls')),
```
Comment (timgraham) [inline tests/urlpatterns/path_urls.py]: It would be better to use or create urls in `tests/urlpatterns` rather than add a dependency on another app.

File: django/urls/resolvers.py (hunk 4 above)
Comment (xavfernandez) [reply django/urls/resolvers.py]: Good point, it belonged in #9323 but is not needed anymore: I removed it.

File: tests/urlpatterns/route_urls.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+from django.conf.urls import include
+from django.urls import path, re_path
+
+from . import views
+
+urlpatterns = [
+    path('articles/2003/', views.empty_view, name='articles-2003'),
+    path('included_urls/', include('urlpatterns_reverse.included_urls')),
+    re_path(r'^regex/(?P<pk>[0-9]+)/$', views.empty_view, name='regex'),
+    path('', include('urlpatterns_reverse.extra_urls')),
```
Comment (timgraham) [inline tests/urlpatterns/route_urls.py]: I think you misunderstood my comment. The idea is that the `urlpatterns_reverse` app shouldn't be used in the `urlpatterns` app.

Comment (xavfernandez) [reply tests/urlpatterns/route_urls.py]: > I think you misunderstood my comment. 

Indeed :crying_cat_face: 

Should I also remove this one: https://github.com/django/django/blob/master/tests/urlpatterns/path_urls.py#L13 ?

Comment (timgraham) [reply tests/urlpatterns/route_urls.py]: Oops, I didn't notice that. It could be done but as a separate commit since it's not related to this issue.

Comment (xavfernandez) [reply tests/urlpatterns/route_urls.py]: https://github.com/django/django/pull/10657/commits/1aaff659437221b104e05fc94b91a4f060c8d8f9 should be that commit

**Comment by xavfernandez** [comment]
Thanks @timgraham for the guidance :)
